### ISSUE TITLE
Eclipse tends to hang when ceylon is installed

### ISSUE THREAD

**Issue by simonthum** [ceylon/ceylon-ide-eclipse#1706]
Quite a few central functions (search & replace, completion) tend to hang Eclipse (fresh Mars 1) "on my machine (TM)".

After much frustration I figure eclipse just didn't get it was done already - classical race involving the the UI thread. Probably something blocks that shouldn't.

What I see is the hourglass and 100% CPU, until I minimize eclipse and bring it up again. Then, the UI thread seems to reapproach the issue and find it done. All fine. It's just that I have to do this quite often.

Minimizing the window is also the only way to get eclipse below 100% CPU, but that may be a different issue, i.e. I did not test specifically if that is ceylon-only.

**Comment by simonthum** [comment]
I'm not sure, but I suspect this is a followup error (from eclipse core jobs):

```
Job found still running after platform shutdown.  Jobs should be canceled by the plugin that scheduled them during shutdown: com.redhat.ceylon.eclipse.code.parse.CeylonParseController$3
```

**Comment by simonthum** [comment]
FWIW I am no longer experiencing the issue. Most noticeable thing I did is switch eclipse to dark theme (which was reeeeally slow) and back again - issue gone.

Unfortunately, this only underlines the brittle nature of race conditions.

**Comment by simonthum** [comment]
Sure enough, it crept back again. I still do not have a definite cause, but I get consistently better usability when unchecking the "Link with editor" button. Perhaps this one does too much on the UI thread.

**Comment by quintesse** [comment]
Bummer :(

**Comment by simonthum** [comment]
Absolutely. A nice symptom I got today was the launch progress dialog still pretending to do "pre-lauch checks" while the unit test part went to green...

**Comment by simonthum** [comment]
OK, since upgrading to 1.2.2 eclipse's default behaviour prevents any form of productivity. However, since the other IDEs aren't there yet, I did two things:

1. Switch off automatic building - this impairs usability but restores performance to a workable degree.

2. In an effort to move my ceylon flow to traditional editors, I measured compiler performance. Lo and behold:

project a (the smaller utility project):

```
real    1m4.716s
user    4m12.784s
sys     0m6.304s
```

The larger (~3x) prototype project:

```
real    3m54.343s
user    15m9.964s
sys     0m16.949s
```

It's not a big project and this seems outsized. Do people have similar times? This is a 4-real-cores i7 north of 2 GHz.

@davidfestal  could you compare the compilation times to those of your projects? It seems to me that long compile times could trigger this whole thing.

Also, it would be very cool to have incremental compilation only when saving, this would make working without automatic build much more comfortable.

**Comment by jvasileff** [comment]
@simonthum that looks way too slow. Perhaps there is something specific in your code that is causing a problem for the typechecker or compiler. For reference, compiling the entire SDK and all test modules on my small 2-core laptop takes only 1m27s (2m34s of user time << 4m12s).

```
jvasileff@lynx:ceylon-sdk$ time ceylon compile -W
Note: Created module test.ceylon.unicode/1.2.2
Note: Created module test.ceylon.numeric/1.2.2
Note: Created module test.ceylon.file/1.2.2
Note: Created module test.ceylon.whole.common/1.2.2
Note: Created module ceylon.test/1.2.2
...

real	1m25.937s
user	2m35.152s
sys	0m5.220s
```

**Comment by simonthum** [comment]
Thanks, at least I know that something's clearly wrong now.  I'll see if profiling gets me further.

**Comment by simonthum** [comment]
OK, for the larger project I'm down to

```
real    0m33.080s
user    2m18.525s
sys     0m1.787s
```

Eclipse feels much more useable now, although I'm less than sure the race conditions cease to exist.

Out of a hunch, I checked my self type declarations and found them to be at odds to those of the langauge module, e.g. ``Comparable<Other>``. I was missing Comparable's `of Other` clause equivalent in my declarations - this did not change anything but compile times and, interestingly, eclipse responsiveness.

So if I'm right, having sources with self types missing the `of` clause is taking the typechecker down so much that it hits hard on eclipse.

The question is whether the observed eclipse behaviour is just a sneak preview of larger projects, and I'm leaving it open for you to figure out. I suppose the typechecker's behaviour is a different matter then.

**Comment by gavinking** [comment]
> Out of a hunch, I checked my self type declarations and found them to be at odds to those of the langauge module, e.g. ``Comparable<Other>``. I was missing Comparable's `of Other` clause equivalent in my declarations - this did not change anything but compile times and, interestingly, eclipse responsiveness.


Wow! That is ... unexpected. I can't even begin to think of an explanation for that...

**Comment by simonthum** [comment]
Same here. I would never have checked those types if I hadn't observed that, besides ceylon upgrades and IDE fuzzing, increased use of abstraction seemed a likely cause. Which dearly reminded me of c++ ;)

**Comment by simonthum** [comment]
**TL;DR: Eclipse.org downloads considered harmful**

To give an update, eclipse is snappy again for me. My solution?

_Use Ooomph to create the smallest-possible eclipse distro you could possibly install ceylon in. _

By wasting a day on investigation, I noticed that most hangs were associated with

 * resource abuse (AERI being the top abuser; 30 megs of .metadata space and 4/5 open file handles)
 * plugins like gradle, m2, mylyn (the top hogs) spinning around themselves for no real reason (like, I may have actually used them to some extent beyond checking their preference pages for expensive-sounding options)

In summary, the (detected) UI freezes were in the order of minutes, now they're typically below a second. Even completion comes up reliably again!

Although ceylon is apparently good in exacerbating these issues, it's not their cause. I recommend to maybe mention a small installation profile on the ceylon IDE page.
